02_knight_and_princess/matveev_alexey/task2.cpp

```diff
@@ -0,0 +1,79 @@
+#include <algorithm>
+#include <array>
+#include <cassert>
+#include <fstream>
+#include <iostream>
+#include <vector>
+
+using VertexId = int;
+using EdgeId = int;
+
+struct Vertex {
+  const VertexId id;
+
+  explicit Vertex(const VertexId& _id) : id(_id) {}
+};
+
+struct Edge {
+  const EdgeId id;
+  const VertexId vertex_id1;
+  const VertexId vertex_id2;
+  Edge(const EdgeId& _id, VertexId _vertex_id1, VertexId _vertex_id2)
+      : id(_id), vertex_id1(_vertex_id1), vertex_id2(_vertex_id2) {}
+};
+
+class Graph {
+ public:
+  bool HasVertex(const VertexId& _vertex_id) {
+    for (const auto& vertex : vertexes_) {
+      if (vertex.id == _vertex_id) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  bool AreConnected(const VertexId& _vertex_id1, const VertexId& _vertex_id2) {
+    assert(HasVertex(_vertex_id1) && "Vertex1 index is out of range");
+    assert(HasVertex(_vertex_id2) && "Vertex2 index is out of range");
+    for (const auto& edge : edges_) {
+      if ((edge.vertex_id1 == _vertex_id1 && edge.vertex_id2 == _vertex_id2) ||
+          (edge.vertex_id1 == _vertex_id2 && edge.vertex_id2 == _vertex_id1)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void AddVertex() { vertexes_.emplace_back(vertex_new_id_++); }
+
+  void AddEdge(const VertexId& _vertex_id1, const VertexId& _vertex_id2) {
+    assert(HasVertex(_vertex_id1) && "Vertex1 index is out of range");
+    assert(HasVertex(_vertex_id2) && "Vertex2 index is out of range");
+    assert(!AreConnected(_vertex_id1, _vertex_id2) &&
+           "These vertexes are already connected");
+    edges_.emplace_back(edge_new_id_++, std::min(_vertex_id1, _vertex_id2),
+                        std::max(_vertex_id1, _vertex_id2));
+  }
+
+ private:
+  std::vector<Vertex> vertexes_;
+  std::vector<Edge> edges_;
+  int vertex_new_id_ = 0, edge_new_id_;
+};
+
+int main() {
+  constexpr int VERTEX_NUMBER = 14, EDGE_NUMBER = 18;
+  const std::vector<std::array<VertexId, 2>> vertex = {
```

Comment:
1. Я тебе вроде предлагал использовать `std::pair`, не получилось?
2. Если ты знаешь размер вектора, то логичней и более эффективно использовать `std::array`
3. `Vertex` - это вершина. Твой же вектор, это что угодно, но не вершина. Нужно переименовать.

```suggestion
  const std::array<std::pair<VertexId, VertexId>, size> connections = {
```

Reply:
Насколько я помню, я пытался использовать и array, и pair, но получилось только с вектором